Add grid reconstruction utilities to reconstruct Grids from NetCDF outputs

tomchor opened the pull request:
Needed to complete https://github.com/CliMA/Oceananigans.jl/issues/3935

This PR allows us to dump grid data (and metadata) into a NetCDF and reconstruct the same grid from it later on:

```julia
ds = NCDataset(filename, "c")
write_grid_reconstruction_data!(ds, original_grid)
close(ds)

reconstructed_grid = reconstruct_grid_from_netcdf(filename)
original_grid == grid # true
```

It also adds tests for the grid reconstruction utilities that already existed, but which do not use NetCDF. (Namely, `constructor_arguments()`.)

navidcy commented on test/test_grid_constructor_arguments.jl:
```diff
@@ -0,0 +1,267 @@
+    # Test coordinate arrays match
+    @test all(reconstructed_grid.λᶠᵃᵃ == original_grid.λᶠᵃᵃ)
+    @test all(reconstructed_grid.λᶜᵃᵃ == original_grid.λᶜᵃᵃ)
+    @test all(reconstructed_grid.φᵃᶠᵃ == original_grid.φᵃᶠᵃ)
+    @test all(reconstructed_grid.φᵃᶜᵃ == original_grid.φᵃᶜᵃ)
```
test z?

navidcy commented on test/test_grid_constructor_arguments.jl:
```diff
@@ -0,0 +1,267 @@
+using Oceananigans
+using Test
```
```suggestion
include("dependencies_for_runtests.jl")
```

tomchor commented:
> Oh sorry, I only noticed it was a draft after I made the suggestions! :(

No worries at all! Comments are always appreciated :)

I'll address them soon.

tomchor commented:
I realized that grid, currently, equality only works properly for non-immersed grids, as it doesn't compare anything related to immersed boundaries. I want this feature to be available for immersed grids as well, as the tests would be much simpler with proper equality testing for immersed grids, so I'm going to open that PR first before testing immersed grids here.

Otherwise this PR is more or less ready.

glwagner commented on ext/OceananigansNCDatasetsExt.jl:
```diff
@@ -709,11 +711,17 @@ function grid_attributes(ibg::ImmersedBoundaryGrid)
 netcdfify(x::NTuple{N, Number}) where N = collect(x)
 netcdfify(x) = string(x)
 
+de_netcdfify_dict_values(dict) = OrderedDict(Symbol(key) => de_netcdfify(value) for (key, value) in dict)
```
```suggestion
materialize_from_netcdf(dict) = OrderedDict(Symbol(key) => de_netcdfify(value) for (key, value) in dict)
```

glwagner commented on ext/OceananigansNCDatasetsExt.jl:
```diff
@@ -676,52 +659,80 @@ function grid_reconstruction_attributes(grid::LatitudeLongitudeGrid)
+function write_grid_reconstruction_data!(ds, grid; array_type=Array{eltype(grid)}, deflatelevel=0)
+    grid_attrs, grid_dims = grid_attributes(grid)
 
+    sorted_grid_attrs = sort(collect(pairs(grid_attrs)), by=first)
```
add a comment about what this is doing?

glwagner reviewed (approved):
Can you put some code into this discussion that illustrates what this does? I presume this allows us to write something like

```julia
grid = construct_grid(filename)
```

when `filename` points to a NetCDF file generated by `NetCDFWriter`.

tomchor commented on src/Grids/rectilinear_grid.jl:
```diff
@@ -387,7 +389,7 @@ cpu_face_constructor_y(grid::YRegularRG) = y_domain(grid)
     arch = architecture(grid)
     FT = eltype(grid)
-    args = Dict(:architecture => arch, :number_type => eltype(grid))
+    args = OrderedDict(:architecture => arch, :number_type => eltype(grid))
```
Added this comment:

`# We use OrderedDict to preserve order of keys. Important for positional arguments since we wanna be able to splat them.`

glwagner commented:
Another thing to consider is that there are a number of additional features for output analysis that we would like to add; notably the ability to inspect the content of a file without having to build `FieldTimeSeries` (ie to see what variables are saved inside). It'd be nice to have a common interface for all file formats for that too.

tomchor commented:
> > > Can you put some code into this discussion that illustrates what this does? I presume this allows us to write something like
> > > ```julia
> > > grid = construct_grid(filename)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > when `filename` points to a NetCDF file generated by `NetCDFWriter`.
> > 
> > 
> > Yes. I added a quick description at the top. Let me know if that's enough. The user current interface is a bit more verbose, but I'm wondering if I should change it to what you used in your example, which is much simpler and still understandable. Let me know your thoughts.
> 
> I think typically fewer and shorter names is better, when we can leverage multiple dispatch / analysis of the arguments to control behavior, the benefit being an interface which is easier to learn and remember. "grid construction" is useful for any file format (not just netcdf) so I'm not sure it's well motivated to append `from_netcdf`.

Makes sense. Just changed it to `reconstruct_grid`. Although I think `construct_grid` might also be good...

> 
> Also we might want something like `reconstruct_grid(::Dataset)` in addition to `reconstruct_grid(::String) since the former is what we need for `FieldTimeSeries`.

That is already what the main method is. The `reconstruct_grid(::String)` method just opens the NCDataset and calls the main method.

> As for `write_grid_reconstruction_data!` --- is it useful to distinguish between "reconstruction data" and the grid itself? Perhaps just `write_grid!` is enough?

I thought it was useful to be more explicit because otherwise it gives the impression that we're writing the grid itself into a file, which (if I understand correctly) can be done with JLD2 but cannot be done with NetCDF. So I thought it was good to make that distinction. But I'm happy to change it if you think it's best. (Also not sure how correct my understanding on JLD2 is...)

tomchor commented:
@glwagner @navidcy  the only tests that are failing are the reactant tests, which don't have anything to do with this PR. Can one of you merge this?
